test/f90_correct/src/use00.f90

```diff
@@ -0,0 +1,29 @@
+module f_precisions
+  integer, parameter, public :: f_double = selected_real_kind(15, 307)
+end module f_precisions
+
+module box
+  use f_precisions, gp=>f_double
+  private
+end module box
+
+module Poisson_Solver
+  use f_precisions
+  integer, parameter, public :: gp=f_double
+end module Poisson_Solver
+
+module module_defs
+  use f_precisions
+  integer, parameter, public :: gp=f_double
+end module module_defs
+
+subroutine PSolver(eh)
+  use module_defs ! This has a public gp
+  use box ! This has a private gp
+  use Poisson_Solver, except_gp => gp ! This has a public gp
+  real(gp), intent(out) :: eh ! Should be able to use gp from module_base
```

Comment:
Given that these are runnable tests, it would be nice to make this a test that actually runs a program which passes iff the Fortran standard is followed in this case. Is it possible to make one of these for the bug that we have fixed? I am not sure it is because I think it might only occur if the private `gp` in `box` and the public `gp` in `Poisson_Solver` alias the same underlying variable. Perhaps a statement that prints out `eh` in `PSolver` to show it is the right precision might work.

In addition, my original bug report contains a second code snippet which is a permutation of the same original code but showing a slightly different bug. This bug is reproducable with a running program as it is the compiler incorrectly selecting the underlying alias of `gp` to use and can be seen in a print statement. A runnable testcase which prints output and checks it against an expected value should be possible for that bug. 

Also - does your patch fix that related bug?

Reply:
I tried to cut the example as short as possible and still keep it failing without the fix and passing with the fix and this is what I came down to. The bug is revealed at compilation time, so I didn't think any particular checks are required, but now that you pointed out the depth of the problem I'll try to add a check for `eh` precision being the same as `gp`'s.

The permuted example failed to compile for me just like the non-permuted one, so I didn't take care of it any more and I must admit I forgot about it. I just checked now that it also compiles fine once the fix is there.

Reply:
I updated the test case code adding a public function allowing us to check the value of `box`'es private `gp` (it's private, so no direct way to access it) and comparing its return value to the publicly available `gp`. The test is passing. I hope I got it right, @RichBarton-Arm ?

Have you managed to reproduce the other issue that you mentioned: different incorrect behaviour with slightly permuted test?